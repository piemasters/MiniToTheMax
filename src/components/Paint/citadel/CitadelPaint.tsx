import React, { FC } from 'react';

import { getSVG } from './getSVG';
import { PaintDetailsCitadel } from '../../../types';

export const CitadelPaint: FC<{ paint: PaintDetailsCitadel }> = ({ paint }) => {
  return (
    <div
      data-testid={`paint_${paint.name}_${paint.type.toString()}`}
      className="inline-block m-1 w-[60px]"
    >
      <div className="flex flex-col justify-center">
        <div className="mb-2">{getSVG(paint)}</div>
        <div className="flex-1 text-[0.6rem] leading-4 text-center max-w-full">
          {paint.name}
        </div>
      </div>
    </div>
  );
};

export default CitadelPaint;
