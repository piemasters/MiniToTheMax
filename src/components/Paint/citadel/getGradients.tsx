import React, { ReactElement } from 'react';

import { textToId } from '../../../util/textToId';
import type { CitadelPaintCategory, PaintGradient } from '../../../types';

export const getGradients = ({
  gradient,
  hex,
  name,
  gloss,
  category,
}: {
  gradient?: PaintGradient[];
  hex: string;
  name: string;
  gloss?: boolean;
  category: CitadelPaintCategory;
}) => {
  if (!gradient) return null;

  const gradients: JSX.Element[] = [];

  if (category === 'Citadel Contrast') {
    gradients.push(
      <linearGradient
        y1="0%"
        x1="100%"
        x2="0%"
        y2="0%"
        id={`${textToId(category)}_gradient_${hex.substring(1, 7)}`}
        key={`${textToId(name)}_${textToId(category)}`}
      >
        {getStopGradients(gradient)}
      </linearGradient>
    );
  } else if (category === 'Citadel Shade' || category === 'Citadel Technical') {
    gradients.push(
      <linearGradient
        y1="100%"
        x1="0%"
        x2="0%"
        y2="0%"
        id={`${textToId(category)}_gradient_${hex.substring(1, 7)}`}
        key={`${textToId(name)}_${textToId(category)}`}
      >
        {getStopGradients(gradient)}
      </linearGradient>
    );
  } else {
    gradients.push(
      <radialGradient
        cx="0"
        cy="0"
        r="1"
        gradientUnits="userSpaceOnUse"
        gradientTransform="translate(25 32) scale(39)"
        id={`${textToId(category)}_gradient_${hex.substring(1, 7)}`}
        key={`${textToId(name)}_${textToId(category)}`}
      >
        {getStopGradients(gradient)}
      </radialGradient>
    );
  }
  if (gloss) {
    gradients.push(
      <radialGradient
        id={`${textToId(category)}_gloss_gradient_${hex.substring(1, 7)}`}
        cx="0"
        cy="0"
        r="1"
        gradientUnits="userSpaceOnUse"
        gradientTransform="translate(14.5226 43.0942) rotate(25.9038) scale(62.7474 27.1872)"
        key={`${textToId(name)}_${textToId(category)}_gloss`}
      >
        <stop stopColor="white" stopOpacity="0.35" />
        <stop offset="1" stopColor="white" stopOpacity="0" />
      </radialGradient>
    );
  }
  return <defs>{gradients}</defs>;
};

export const getStopGradients = (gradient: PaintGradient[]) => {
  const gradients: ReactElement[] = [];
  if (gradient) {
    for (const g of gradient) {
      gradients.push(
        <stop
          offset={`${g.offset}%`}
          stopColor={g.color}
          stopOpacity={`${g.opacity || 1}`}
          key={g.offset}
        />
      );
    }
    return gradients;
  }
};
