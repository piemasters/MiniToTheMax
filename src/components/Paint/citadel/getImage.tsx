import { CitadelTechnicalPaintImages } from '../../../types';
import AgrellanBadland from '../../../../content/assets/svg/paints/technicalAgrellanBadland.svg';
import AgrellanEarth from '../../../../content/assets/svg/paints/technicalAgrellanEarth.svg';
import ArmageddonDunes from '../../../../content/assets/svg/paints/technicalArmageddonDunes.svg';
import ArmageddonDust from '../../../../content/assets/svg/paints/technicalArmageddonDust.svg';
import Astrogranite from '../../../../content/assets/svg/paints/technicalAstrogranite.svg';
import AstrograniteDebris from '../../../../content/assets/svg/paints/technicalAstrograniteDebris.svg';
import MartianIroncrust from '../../../../content/assets/svg/paints/technicalMartianIroncrust.svg';
import MartianIronearth from '../../../../content/assets/svg/paints/technicalMartianIronearth.svg';
import MordantEarth from '../../../../content/assets/svg/paints/technicalMordantEarth.svg';
import StirlandBattlemire from '../../../../content/assets/svg/paints/technicalStirlandBattlemire.svg';
import StirlandMud from '../../../../content/assets/svg/paints/technicalStirlandMud.svg';
import ValhallanBlizzard from '../../../../content/assets/svg/paints/technicalValhallanBlizzard.svg';

const paintImages = {
  AgrellanBadland: <AgrellanBadland />,
  AgrellanEarth: <AgrellanEarth />,
  ArmageddonDunes: <ArmageddonDunes />,
  ArmageddonDust: <ArmageddonDust />,
  Astrogranite: <Astrogranite />,
  AstrograniteDebris: <AstrograniteDebris />,
  MartianIroncrust: <MartianIroncrust />,
  MartianIronearth: <MartianIronearth />,
  MordantEarth: <MordantEarth />,
  StirlandBattlemire: <StirlandBattlemire />,
  StirlandMud: <StirlandMud />,
  ValhallanBlizzard: <ValhallanBlizzard />,
};

export const getImage = (name: CitadelTechnicalPaintImages) => {
  return paintImages[name];
};
