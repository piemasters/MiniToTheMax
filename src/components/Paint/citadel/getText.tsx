import React from 'react';

import { PaintGradient } from '../../../types';
import { getColorByBgColor } from './getColorByBgColor';

export const getText = ({
  type,
  hex,
  gradient,
}: {
  type: string;
  hex: string;
  gradient?: PaintGradient[];
}) => {
  if (type === 'dry') {
    return (
      <g
        transform="translate(16 48), scale(.5 .5)"
        fill={gradient ? '#292929' : getColorByBgColor(hex)}
      >
        <g clipPath="url(#clip_dry)">
          <path d="M3.897.074c4.985 0 7.6 2.42 7.6 6.998 0 5.062-2.633 7.798-7.565 7.798-1.387-.019-2.686-.168-3.932-.559V.67C1.018.298 2.334.074 3.897.074zm.087 12.73c3.511 0 5.231-1.824 5.231-5.583 0-3.424-1.702-5.1-5.16-5.1-.527 0-1.159.038-1.826.15v10.365a9.923 9.923 0 0 0 1.755.168zM21.765 15l-5.248-7.872c2.633 0 3.914-.763 3.914-2.457 0-1.693-1.036-2.568-3.072-2.568-.509 0-.965.019-1.456.112V14.59c-.387.093-.738.13-1.141.13-.386 0-.72-.037-1.071-.13V.428C15.06.205 16.254.074 17.289.074c3.581 0 5.477 1.582 5.477 4.355 0 1.936-1.053 3.276-3.037 3.815l3.985 5.286c-.457.744-1.106 1.228-1.949 1.47zM29.295 6.774L33 0c.86.112 1.492.521 2.001 1.191l-4.599 7.612v5.788c-.333.093-.72.111-1.088.111-.386 0-.737-.018-1.123-.111V8.803l-4.6-7.612C24.1.502 24.75.111 25.593 0l3.703 6.774z" />
        </g>
        <defs>
          <clipPath id="clip_dry">
            <path d="M0 0h35v15H0z" />
          </clipPath>
        </defs>
      </g>
    );
  } else if (type === 'air') {
    return (
      <g
        transform="translate(16 48), scale(.5 .5)"
        fill={gradient ? '#292929' : getColorByBgColor(hex)}
      >
        <g clipPath="url(#clip_air)">
          <path d="M11.2 15.85l-1.4-3.26H3.3c-.5 1.3-.76 1.94-1.25 3.24-.79 0-1.48-.33-2.05-1L6.42 0c2.26 4.94 4.52 9.88 6.8 14.8-.55.66-1.24 1-2.03 1.04zM6.41 4.77c-.78 1.91-1.58 3.83-2.36 5.77h4.89L6.42 4.77zM15.82 15.67c-.43 0-.78-.02-1.1-.12V.68a4.58 4.58 0 0 1 2.22 0v14.87c-.32.1-.7.12-1.12.12zM28.02 16L22.7 7.79c1.55.1 3.99-.37 3.97-2.56-.02-1.74-1.05-2.67-3.11-2.67-.52 0-.98.02-1.48.11v12.88a4.62 4.62 0 0 1-2.24 0V.81c1.39-.23 2.63-.34 3.65-.36 2.52-.08 5.55 1.33 5.55 4.53 0 2-1.32 3.54-3.08 3.97 1.35 1.84 2.69 3.68 4.04 5.5A3.41 3.41 0 0 1 28.02 16z" />
        </g>
        <defs>
          <clipPath id="clip_air">
            <path d="M0 0h30v16H0z" />
          </clipPath>
        </defs>
      </g>
    );
  } else {
    return null;
  }
};
