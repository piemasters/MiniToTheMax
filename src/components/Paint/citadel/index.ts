export * from './getGloss';
export * from './getGradients';
export * from './getShape';
export * from './getText';
