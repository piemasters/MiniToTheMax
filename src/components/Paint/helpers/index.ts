export * from './getColorByBgColor';
export * from './getGloss';
export * from './getGradients';
export * from './getShape';
export * from './getStopGradients';
export * from './getSVG';
export * from './getText';
