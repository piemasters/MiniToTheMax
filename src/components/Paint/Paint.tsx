import React, { FC } from 'react';

import type {
  PaintDetails,
  CitadelPaintCategory,
  CitadelTechnicalPaintImages,
  VallejoPaintCategory,
} from '../../types';
import VallejoPaint from './vallejo/VallejoPaint';
import CitadelPaint from './citadel/CitadelPaint';

export const Paint: FC<{ paint: PaintDetails }> = ({ paint }) => {
  if (paint.company === 'Vallejo') {
    return (
      <VallejoPaint
        name={paint.name}
        type={paint.type}
        color={paint.color}
        hex={paint.hex}
        gradient={paint.gradient}
        availability={paint.availability}
        company={paint.company}
        category={paint.category as VallejoPaintCategory}
        stroke={paint.stroke}
        gloss={paint.gloss}
        img={paint.img}
        number={paint.number || ''}
      />
    );
  }
  return (
    <CitadelPaint
      name={paint.name}
      type={paint.type}
      color={paint.color}
      hex={paint.hex}
      gradient={paint.gradient}
      availability={paint.availability}
      company={paint.company as 'Citadel'}
      category={paint.category as CitadelPaintCategory}
      stroke={paint.stroke}
      gloss={paint.gloss}
      img={paint.img as CitadelTechnicalPaintImages | undefined}
    />
  );
};

export default Paint;
