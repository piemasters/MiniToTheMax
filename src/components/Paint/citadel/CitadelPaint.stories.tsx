import type { Meta, StoryObj } from '@storybook/react-vite';

import { CitadelPaint } from './CitadelPaint';
import { PaintDetailsCitadel, PaintType } from '../../../types';

const meta: Meta<PaintDetailsCitadel> = {
  component: CitadelPaint,
  argTypes: {
    name: {
      description: 'The name of the paint color',
    },
    type: {
      description: 'The type of the paint',
      control: 'check',
      options: [
        'air',
        'core',
        'dry',
        'fluorescent',
        'ink',
        'metallic',
        'one-coat',
        'spray',
        'technical',
      ] as PaintType[],
    },
    color: {
      description: 'The color of the paint',
      control: 'radio',
      options: [
        'black',
        'blue',
        'bone',
        'brass',
        'bronze',
        'brown',
        'clear',
        'copper',
        'flesh',
        'gold',
        'green',
        'grey',
        'orange',
        'pink',
        'purple',
        'red',
        'silver',
        'turquoise',
        'white',
        'yellow',
      ],
    },
    hex: {
      description: 'The hex code of the paint color',
      control: 'color',
    },
    gradient: {
      description: 'The gradient of the paint color',
      control: 'object',
    },
    availability: {
      description: 'The availability status of the paint',
      control: 'radio',
      options: ['available', 'discontinued'],
    },
    company: {
      description: 'The company that produces the paint',
      control: 'radio',
      options: ['Citadel'],
    },
    category: {
      description: 'The category of the paint',
      control: 'radio',
      options: [
        'Citadel Air',
        'Citadel Base',
        'Citadel Contrast',
        'Citadel Dry',
        'Citadel Layer',
        'Citadel Shade',
        'Citadel Spray',
        'Citadel Technical',
      ],
    },
    stroke: {
      description: 'The hex code of the paint color',
      control: 'color',
    },
    gloss: {
      description: 'Whether the paint has a gloss effect',
      control: 'boolean',
    },
    img: {
      description: 'The image name of the paint',
      control: 'radio',
      options: [
        undefined,
        'AgrellanBadland',
        'AgrellanEarth',
        'ArmageddonDunes',
        'ArmageddonDust',
        'Astrogranite',
        'AstrograniteDebris',
        'MartianIroncrust',
        'MartianIronearth',
        'MordantEarth',
        'StirlandBattlemire',
        'StirlandMud',
        'ValhallanBlizzard',
      ],
    },
  },
};

export const Default: Story = {
  args: {
    name: 'Averland Sunset',
    category: 'Citadel Base',
    type: ['core'],
    color: 'yellow',
    hex: '#FBB81C',
    gradient: [
      { offset: 0, color: 'rgb(252, 235, 211)' },
      { offset: 50, color: 'rgb(191, 117, 63)' },
      { offset: 100, color: 'rgb(61, 7, 0)' },
    ],
    stroke: '#000000',
    availability: 'available',
    company: 'Citadel',
    gloss: false,
    img: undefined,
  },
};

export default meta;
type Story = StoryObj<PaintDetailsCitadel>;
