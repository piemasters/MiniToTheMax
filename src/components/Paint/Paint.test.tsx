import { getByTestId, render } from '@testing-library/react';

import { Paint } from './Paint';
import { baseRetributorArmour } from '../../data/paints/citadel/base';

describe('Paint', () => {
  test('renders correctly', () => {
    const { container } = render(<Paint paint={baseRetributorArmour} />);

    expect(getByTestId(container as HTMLElement, 'paint')).toBeTruthy();
    expect(container).toMatchSnapshot();
  });
});
