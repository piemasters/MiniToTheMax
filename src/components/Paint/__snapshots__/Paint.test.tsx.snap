// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Paint renders correctly 1`] = `
<div>
  <div
    class="inline-block m-1 w-[60px]"
    data-testid="paint"
  >
    <div
      class="flex flex-col justify-center"
    >
      <div
        class="mb-2"
      >
        <svg
          height="71"
          viewBox="-1 0 52 62"
          width="57"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M34.9927 3.31021C34.0377 1.87791 29.0583 0.923096 24.4199 0.923096C19.7814 0.923096 16.0844 2.00071 14.3927 3.31021C8.73112 4.94709 5.1841 8.56187 4.63841 11.5628C4.09271 14.5638 0 46.0055 0 48.7337C0 51.4618 4.50198 61.6923 25.3066 61.6923C46.1112 61.6923 49.9311 51.3254 49.9993 48.7337C50.0675 46.142 45.0198 10.7444 45.0198 10.7444C44.2695 5.76553 35.6748 3.37842 34.9927 3.31021Z"
            fill="url(#base_gradient_89571D)"
          />
          <defs>
            <radialgradient
              cx="0"
              cy="0"
              gradientTransform="translate(25 32) scale(39)"
              gradientUnits="userSpaceOnUse"
              id="base_gradient_89571D"
              r="1"
            >
              <stop
                offset="0%"
                stop-color="rgb(255, 255, 255)"
                stop-opacity="1"
              />
              <stop
                offset="50%"
                stop-color="rgb(235, 184, 84)"
                stop-opacity="1"
              />
              <stop
                offset="100%"
                stop-color="rgb(137, 87, 29)"
                stop-opacity="1"
              />
            </radialgradient>
          </defs>
        </svg>
      </div>
      <div
        class="flex-1 text-[0.6rem] leading-4 text-center max-w-full"
      >
        Retributor Armour
      </div>
    </div>
  </div>
</div>
`;
